File: examples/simple-examples-java/src/main/java/ai/koog/agents/example/calculator/Calculator.java
package ai.koog.agents.example.calculator;

import static ai.koog.prompt.executor.llms.all.SimplePromptExecutors.simpleOllamaAIExecutor;
import static ai.koog.prompt.executor.llms.all.SimplePromptExecutors.simpleOpenAIExecutor;

import ai.koog.agents.core.agent.AIAgent;
import ai.koog.agents.core.agent.entity.AIAgentEdge;
import ai.koog.agents.core.agent.entity.AIAgentNode;
import ai.koog.agents.core.agent.entity.AIAgentNodeBase;
import ai.koog.agents.core.agent.entity.GraphStrategyBuilder;
import ai.koog.agents.core.agent.entity.TypedGraphStrategyBuilder;
import ai.koog.agents.core.dsl.extension.HistoryCompressionStrategy;
import ai.koog.agents.core.dsl.extension.ToolCalls;
import ai.koog.agents.core.tools.ToolRegistry;
import ai.koog.agents.example.ApiKeyService;
import ai.koog.agents.ext.tool.AskUser;
import ai.koog.agents.ext.tool.SayToUser;
import ai.koog.agents.features.eventHandler.feature.EventHandler;
import ai.koog.prompt.executor.clients.openai.OpenAIModels.Chat;
import ai.koog.prompt.executor.model.PromptExecutor;
import ai.koog.prompt.executor.ollama.client.OllamaModels.Meta;
import ai.koog.prompt.llm.LLModel;
import ai.koog.prompt.message.Message;

/**
 * Demonstrates how to build a graph-based calculator agent in Java using the Koog Java API:
 * <ul>
 *   <li>Defining tools via {@link CalculatorTools} (a {@code ToolSet} implementation)</li>
 *   <li>Building a graph strategy with multiple nodes and typed edges</li>
 *   <li>Handling agent events via the {@code EventHandler} feature</li>
 * </ul>
 *
 * <p>Usage:
 * <ul>
 *   <li>Run with OpenAI GPT-4o (requires {@code OPENAI_API_KEY}): {@code ./gradlew runExampleCalculator}</li>
 *   <li>Run with local Ollama Llama 3.2: {@code ./gradlew runExampleCalculatorLocal}</li>
 * </ul>
 */
public class Calculator {

    private static final int MAX_TOKENS = 1000;

    enum RunMode {
        LOCAL_LLAMA_3_2,
        OPEN_AI_GPT_4o
    }

    public static void main(String[] args) {
        RunMode runMode = args.length > 0 && "local".equalsIgnoreCase(args[0])
            ? RunMode.LOCAL_LLAMA_3_2
            : RunMode.OPEN_AI_GPT_4o;
        runCalculatorExample(runMode);
    }

    private static void runCalculatorExample(RunMode runMode) {
        try (PromptExecutor executor = chooseExecutor(runMode)) {
            LLModel model = chooseModel(runMode);

            // 1. Build the tool registry with calculator tools
            ToolRegistry toolRegistry = ToolRegistry.builder()
                .tool(AskUser.INSTANCE)
                .tool(SayToUser.INSTANCE)
                .tools(new CalculatorTools())
                .build();

            // 2. Build the agent with an inline graph strategy
            AIAgent<String, String> agent = AIAgent.builder()
                .promptExecutor(executor)
                .llmModel(model)
                .systemPrompt("You are a calculator.")
                .maxIterations(50)
                .toolRegistry(toolRegistry)
                .<String, String>graphStrategy("calculator", (GraphStrategyBuilder builder) -> {
                    TypedGraphStrategyBuilder<String, String> graph =
                        builder.withInput(String.class).withOutput(String.class);

                    // --- Node definitions ---
                    AIAgentNodeBase<Message.User, Message.Assistant> nodeCallLLM =
                        AIAgentNode.llmRequest("callLLM");

                    AIAgentNodeBase<ToolCalls, Message.User> nodeExecuteTools =
                        AIAgentNode.executeTools("executeTools");

                    AIAgentNodeBase<Message.User, Message.Assistant> nodeSendResults =
                        AIAgentNode.llmRequest("sendToolResults");

                    AIAgentNodeBase<Message.User, Message.User> nodeCompress =
                        AIAgentNode.llmCompressHistory("compressHistory")
                            .withInput(Message.User.class)
                            .compressionStrategy(HistoryCompressionStrategy.WholeHistory)
                            .build();

                    // --- Edge definitions ---

                    // start → callLLM (wrap the agent's String input as a user message)
                    graph.edge(
                        AIAgentEdge.builder().from(graph.nodeStart).to(nodeCallLLM)
                            .asUserMessage()
                            .build()
                    );

                    // callLLM → finish (when LLM returns a text response)
                    graph.edge(
                        AIAgentEdge.builder().from(nodeCallLLM).to(graph.nodeFinish)
                            .onTextMessage()
                            .build()
                    );

                    // callLLM → executeTools (when LLM returns tool calls)
                    graph.edge(
                        AIAgentEdge.builder().from(nodeCallLLM).to(nodeExecuteTools)
                            .onToolCalls()
                            .build()
                    );

                    // executeTools → compress (when token count exceeds threshold)
                    graph.edge(
                        AIAgentEdge.builder().from(nodeExecuteTools).to(nodeCompress)
                            .onCondition((output, ctx) ->
                                ctx.getLlm().readSession(session ->
                                    session.getPrompt().latestTokenUsage() > MAX_TOKENS))
                            .build()
                    );

                    // compress → sendResults
                    graph.edge(nodeCompress, nodeSendResults);

                    // executeTools → sendResults (when token count is within threshold)
                    graph.edge(
                        AIAgentEdge.builder().from(nodeExecuteTools).to(nodeSendResults)
                            .onCondition((output, ctx) ->
                                ctx.getLlm().readSession(session ->
                                    session.getPrompt().latestTokenUsage() <= MAX_TOKENS))
                            .build()
                    );

                    // sendResults → executeTools (when LLM requests more tool calls)
                    graph.edge(
                        AIAgentEdge.builder().from(nodeSendResults).to(nodeExecuteTools)
                            .onToolCalls()
                            .build()
                    );

                    // sendResults → finish (when LLM returns a final text response)
                    graph.edge(
                        AIAgentEdge.builder().from(nodeSendResults).to(graph.nodeFinish)
                            .onTextMessage()
                            .build()
                    );

                    return graph.build();
                })
                .install(EventHandler.Feature, config -> {
                    config.onToolCallStarting(ctx ->
                        System.out.println("Tool called: " + ctx.getToolName()
                                           + ", args: " + ctx.getToolArgs()));
                    config.onAgentExecutionFailed(ctx ->
                        System.out.println("An error occurred: " + ctx.getError().getMessage()));
                    config.onAgentCompleted(ctx ->
                        System.out.println("Result: " + ctx.getResult()));
                })
                .build();

            // 3. Run the agent (blocking call via @JvmName("run"))
            String result = agent.run("(10 + 20) * (5 + 5) / (2 - 11)");
            System.out.println("Agent result: " + result);
        } catch (Exception e) {
            System.err.println("Calculator example failed in mode: " + runMode);
            if (runMode == RunMode.LOCAL_LLAMA_3_2) {
                System.err.println(
                    "Check that Ollama is running at http://localhost:11434 and model llama3.2 is available.");
            } else {
                System.err.println("Check that OPENAI_API_KEY is set and valid.");
            }
            throw new RuntimeException("Unable to run Calculator example", e);
        }
    }

    private static PromptExecutor chooseExecutor(RunMode mode) {
        return switch (mode) {
            case LOCAL_LLAMA_3_2 -> simpleOllamaAIExecutor("http://localhost:11434");
            case OPEN_AI_GPT_4o -> simpleOpenAIExecutor(ApiKeyService.getOpenAIApiKey());
        };
    }

    private static LLModel chooseModel(RunMode mode) {
        return switch (mode) {
            case LOCAL_LLAMA_3_2 -> Meta.LLAMA_3_2;
            case OPEN_AI_GPT_4o -> Chat.GPT4o;
        };
    }
}

File: examples/simple-examples/src/main/kotlin/ai/koog/agents/example/calculator/Calculator.kt
package ai.koog.agents.example.calculator

import ai.koog.agents.core.agent.AIAgent
import ai.koog.agents.core.agent.config.AIAgentConfig
import ai.koog.agents.core.tools.ToolRegistry
import ai.koog.agents.example.ApiKeyService
import ai.koog.agents.ext.tool.AskUser
import ai.koog.agents.ext.tool.SayToUser
import ai.koog.agents.features.eventHandler.feature.handleEvents
import ai.koog.prompt.dsl.prompt
import ai.koog.prompt.executor.clients.openai.OpenAIModels
import ai.koog.prompt.executor.llms.all.simpleOllamaAIExecutor
import ai.koog.prompt.executor.llms.all.simpleOpenAIExecutor
import ai.koog.prompt.executor.model.PromptExecutor
import ai.koog.prompt.executor.ollama.client.OllamaModels
import ai.koog.prompt.llm.LLModel

private enum class RunMode {
    LOCAL_LLAMA_3_2,
    OPEN_AI_GPT_4o
}

/**
 * Demonstrates how to build a graph-based calculator agent in Kotlin using Koog APIs:
 * - Defining tools via [CalculatorTools]
 * - Building a graph strategy with multiple nodes and typed edges
 * - Handling agent events via `handleEvents`
 *
 * Usage:
 * - Run with OpenAI GPT-4o (requires `OPENAI_API_KEY`): `./gradlew runExampleCalculator`
 * - Run with local Ollama Llama 3.2: `./gradlew runExampleCalculatorLocal`
 */
suspend fun main(args: Array<String>) {
    val runMode = when (args.firstOrNull()) {
        "local" -> RunMode.LOCAL_LLAMA_3_2
        else -> RunMode.OPEN_AI_GPT_4o
    }

    runCalculatorExample(runMode)
}

private suspend fun runCalculatorExample(runMode: RunMode) {
    val executor = chooseExecutor(runMode)
    val model = chooseModel(runMode)

    // Create a tool registry with calculator tools
    val toolRegistry = ToolRegistry {
        // Special tool, required with this type of agent.
        tool(AskUser)
        tool(SayToUser)
        tools(CalculatorTools().asTools())
    }

    // Create agent config with proper prompt
    val agentConfig = AIAgentConfig(
        prompt = prompt("test") {
            system("You are a calculator.")
        },
        model = model,
        maxAgentIterations = 50
    )

    executor.use { promptExecutor ->
        // Create the runner
        val agent = AIAgent(
            promptExecutor = promptExecutor,
            strategy = CalculatorStrategy.strategy,
            agentConfig = agentConfig,
            toolRegistry = toolRegistry
        ) {
            handleEvents {
                onToolCallStarting { eventContext ->
                    println("Tool called: tool ${eventContext.toolName}, args ${eventContext.toolArgs}")
                }

                onAgentExecutionFailed { eventContext ->
                    println(
                        "An error occurred: ${eventContext.error.message}\n${eventContext.error.stackTraceToString()}"
                    )
                }

                onAgentCompleted { eventContext ->
                    println("Result: ${eventContext.result}")
                }
            }
        }

        val result = agent.run("(10 + 20) * (5 + 5) / (2 - 11)")
        println("Agent result: $result")
    }
}

private fun chooseExecutor(mode: RunMode): PromptExecutor = when (mode) {
    RunMode.LOCAL_LLAMA_3_2 -> simpleOllamaAIExecutor("http://localhost:11434")
    RunMode.OPEN_AI_GPT_4o -> simpleOpenAIExecutor(ApiKeyService.openAIApiKey)
}

private fun chooseModel(mode: RunMode): LLModel = when (mode) {
    RunMode.LOCAL_LLAMA_3_2 -> OllamaModels.Meta.LLAMA_3_2
    RunMode.OPEN_AI_GPT_4o -> OpenAIModels.Chat.GPT4o
}
